### PR TITLE
[fix](rowset meta) Fix rowset meta size relation

### PR DESCRIPTION
**Pull request by Yukang-Lian**
## Proposed changes

Issue #40879 step 2

make rowset total size = rowset data size + rowset index size

### CODE REVIEW

**Comment by Yukang-Lian** [comment]
Tests will be added to verify the correction of this PR.

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by gavinchou — APPROVED** [review]

File: cloud/src/meta-service/meta_service_txn.cpp [hunk 1]
```diff
@@ -1161,7 +1161,7 @@ void commit_txn_immediately(
 
             // Accumulate affected rows
             auto& stats = tablet_stats[tablet_id];
-            stats.data_size += i.data_disk_size();
+            stats.data_size += i.total_disk_size();
```
Comment (gavinchou) [inline cloud/src/meta-service/meta_service_txn.cpp:1164]: what about BE with older version?

**Review by gavinchou — CHANGES_REQUESTED** [review]
pls add  regression test to ensure the correctness from this PR on

**Comment by zhannngchen** [comment]
> pls add regression test to ensure the correctness from this PR on

+1

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run p0

**Review by gavinchou — DISMISSED** [review]

**Review by bobhan1 — APPROVED** [review]

**Review by w41ter — DISMISSED** [review]

**Comment by Yukang-Lian** [comment]
run buildall

**Review by gavinchou — DISMISSED** [review]

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run cloudut

File: be/src/olap/rowset/rowset_meta.cpp [hunk 2]
```diff
@@ -225,6 +225,7 @@ void RowsetMeta::merge_rowset_meta(const RowsetMeta& other) {
     set_data_disk_size(data_disk_size() + other.data_disk_size());
     set_total_disk_size(total_disk_size() + other.total_disk_size());
     set_index_disk_size(index_disk_size() + other.index_disk_size());
+    set_total_disk_size(data_disk_size() + index_disk_size());
```
Comment (dataroaring) [inline be/src/olap/rowset/rowset_meta.cpp:228]: It is confusing to set total_disk_size twice.

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 3]
```diff
@@ -81,7 +81,7 @@ void build_rowset_meta_with_spec_field(RowsetMeta& rowset_meta,
                                        const RowsetMeta& spec_rowset_meta) {
     rowset_meta.set_num_rows(spec_rowset_meta.num_rows());
     rowset_meta.set_total_disk_size(spec_rowset_meta.total_disk_size());
-    rowset_meta.set_data_disk_size(spec_rowset_meta.total_disk_size());
+    rowset_meta.set_data_disk_size(spec_rowset_meta.data_disk_size());
```
Comment (dataroaring) [inline be/src/olap/rowset/beta_rowset_writer.cpp:84]: We should check impact.

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 4]
```diff
@@ -830,7 +830,8 @@ Status BaseBetaRowsetWriter::_build_rowset_meta(RowsetMeta* rowset_meta, bool ch
 
     rowset_meta->set_num_segments(segment_num);
     rowset_meta->set_num_rows(num_rows_written + _num_rows_written);
-    rowset_meta->set_total_disk_size(total_data_size + _total_data_size);
+    rowset_meta->set_total_disk_size(total_data_size + _total_data_size + total_index_size +
+                                     _total_index_size);
     rowset_meta->set_data_disk_size(total_data_size + _total_data_size);
     rowset_meta->set_index_disk_size(total_index_size + _total_index_size);
```
Comment (dataroaring) [inline be/src/olap/rowset/beta_rowset_writer.cpp:836]: rowset_meta->set_total_disk_size(rowset_meta->get_data_disk_size() + rowset_meta->get_index_disk_size());

File: be/src/olap/rowset/rowset_meta.cpp (hunk 2 above)
Comment (Yukang-Lian) [reply be/src/olap/rowset/rowset_meta.cpp:228]: Yes, it needs just one set operation.

File: be/src/olap/rowset/beta_rowset_writer.cpp (hunk 4 above)
Comment (Yukang-Lian) [reply be/src/olap/rowset/beta_rowset_writer.cpp:836]: OK

File: be/src/olap/rowset/beta_rowset_writer.cpp (hunk 3 above)
Comment (Yukang-Lian) [reply be/src/olap/rowset/beta_rowset_writer.cpp:84]: The impact has already been checked. In old version, total_disk_size=data_disk_size. Index size is not considered at all. This PR we change it to total_disk_size=data_disk_size+index_size.

File: be/src/olap/rowset/beta_rowset_writer.cpp [hunk 5]
```diff
@@ -830,7 +830,8 @@ Status BaseBetaRowsetWriter::_build_rowset_meta(RowsetMeta* rowset_meta, bool ch
 
     rowset_meta->set_num_segments(segment_num);
     rowset_meta->set_num_rows(num_rows_written + _num_rows_written);
-    rowset_meta->set_total_disk_size(total_data_size + _total_data_size);
+    rowset_meta->set_total_disk_size(total_data_size + _total_data_size + total_index_size +
```
Comment (csun5285) [inline be/src/olap/rowset/beta_rowset_writer.cpp:833]: If you want to use the formula ·total size = column data size + index size·, you need to modify the data size and index size statistics in the segment creator, vertical segment writer, and vertical beta rowset writer.

**Review by csun5285 — CHANGES_REQUESTED** [review]

File: be/src/olap/rowset/beta_rowset_writer.cpp (hunk 5 above)
Comment (csun5285) [reply be/src/olap/rowset/beta_rowset_writer.cpp:833]: Currently, `data size = segment file size + inverted index file size`, `total size = data size`

**Review by gavinchou — CHANGES_REQUESTED** [review]

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by gavinchou — DISMISSED** [review]

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run buildall

**Review by gavinchou — DISMISSED** [review]

**Comment by Yukang-Lian** [comment]
run buildall

**Comment by Yukang-Lian** [comment]
run p0

**Review by gavinchou — APPROVED** [review]

**Review by gavinchou — APPROVED** [review]

**Comment by Yukang-Lian** [comment]
run cloud_p0

**Comment by Yukang-Lian** [comment]
run external

**Comment by Yukang-Lian** [comment]
run p0

File: be/src/cloud/cloud_meta_mgr.cpp [hunk 6]
```diff
@@ -1125,4 +1128,124 @@ Status CloudMetaMgr::remove_old_version_delete_bitmap(
...
+    return total_segment_size;
+}
+
+int64_t CloudMetaMgr::get_inverted_index_file_szie(const RowsetMeta& rs_meta) {
+    int64_t total_inverted_index_size = 0;
+    const auto fs = const_cast<RowsetMeta&>(rs_meta).fs();
+    if (!fs) {
+        LOG(WARNING) << "get fs failed, resource_id={}" << rs_meta.resource_id();
+    }
+    if (rs_meta.tablet_schema()->get_inverted_index_storage_format() ==
```
Comment (csun5285) [inline be/src/cloud/cloud_meta_mgr.cpp:1188]: if (!rs_meta.tablet_schema()->has_inverted_index()) {
    return 0;
}

File: be/src/olap/tablet.cpp [hunk 7]
```diff
@@ -2737,4 +2741,120 @@ void Tablet::clear_cache() {
...
+    return total_segment_size;
+}
+
+int64_t Tablet::get_inverted_index_file_szie(const RowsetMetaSharedPtr& rs_meta) {
+    const auto& fs = rs_meta->fs();
+    if (!fs) {
+        LOG(WARNING) << "get fs failed, resource_id={}" << rs_meta->resource_id();
+    }
+    int64_t total_inverted_index_size = 0;
+
```
Comment (csun5285) [inline be/src/olap/tablet.cpp:2808]: if (!rs_meta->tablet_schema()->has_inverted_index()) {
    return 0;
}

File: be/src/cloud/cloud_meta_mgr.cpp [hunk 8]
```diff
@@ -1125,4 +1128,124 @@ Status CloudMetaMgr::remove_old_version_delete_bitmap(
...
+                        LOG(WARNING)
+                                << "cloud table size correctness check get inverted index v1 "
+                                   "size failed! msg:"
+                                << st.msg() << ", inverted index path:" << inverted_index_file_path;
+                    }
+                }
+                total_inverted_index_size += file_size;
+            }
+        }
+    } else {
```
Comment (csun5285) [inline be/src/cloud/cloud_meta_mgr.cpp:1223]: if (rs_meta.tablet_schema()->has_inverted_index())

File: be/src/runtime/load_stream_writer.cpp [hunk 9]
```diff
@@ -201,7 +201,7 @@ Status LoadStreamWriter::add_segment(uint32_t segid, const SegmentStatistics& st
     }
 
     DBUG_EXECUTE_IF("LoadStreamWriter.add_segment.size_not_match", { segment_file_size++; });
-    if (segment_file_size + inverted_file_size != stat.data_size) {
+    if (segment_file_size != stat.data_size) {
```
Comment (csun5285) [inline be/src/runtime/load_stream_writer.cpp:204]: segment_file_size != stat.data_size || inverted_file_size != stat.index_size

File: be/src/olap/tablet.cpp [hunk 10]
```diff
@@ -2737,4 +2741,120 @@ void Tablet::clear_cache() {
...
+                                 << ", rowset id:" << rs_meta->rowset_id()
+                                 << ", table size correctness check get inverted index v1 "
+                                    "size failed! msg:"
+                                 << st.to_string()
+                                 << ", inverted index path:" << inverted_index_file_path;
+                }
+                total_inverted_index_size += file_size;
+            }
+        }
+    } else {
```
Comment (csun5285) [inline be/src/olap/tablet.cpp:2838]: if (rs_meta->tablet_schema()->has_inverted_index())

File: be/src/cloud/cloud_meta_mgr.cpp (hunk 8 above)
Comment (csun5285) [inline be/src/cloud/cloud_meta_mgr.cpp:1223]: if (rs_meta.tablet_schema()->has_inverted_index())

File: be/src/olap/tablet.cpp (hunk 10 above)
Comment (csun5285) [inline be/src/olap/tablet.cpp:2838]: if (rs_meta->tablet_schema()->has_inverted_index())

File: be/src/runtime/load_stream_writer.cpp (hunk 9 above)
Comment (csun5285) [inline be/src/runtime/load_stream_writer.cpp:204]: if (segment_file_size != stat.data_size || inverted_file_size != stat.index_size)

**Pull request merged**